### PR TITLE
♻️ Refactor API-keys service

### PR DESCRIPTION
**Pull request by giancarloromeo**
## What do these changes do?
[ReDoc](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/ITISFoundation/osparc-simcore/90295ecab257275c863e82297e39a985fabd1147/services/web/server/src/simcore_service_webserver/api/v0/openapi.yaml) 

The `API Keys` WebServer's plugin needed some changes (see related issues) and a refactoring.

Please, have a look at the new generated ReDoc... especially in `/auth/api-keys` endpoints.

## Related issue/s

- Resolves: #4959 
- Partially resolves: #2280 
- Resolves: #5022
- Resolves: #4920

## Dev-ops checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

**Review by odeimaiz — APPROVED** [review]
Looks good to me

File: services/web/server/src/simcore_service_webserver/api_keys/_service.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+async def create_api_key(
+    app: web.Application,
+    *,
+    user_id: UserID,
+    product_name: ProductName,
+    display_name=str,
+    expiration=dt.timedelta,
+) -> ApiKey:
+    # generate key and secret
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/api_keys/_service.py:37]: redundant doc ;-)

File: packages/models-library/src/models_library/api_schemas_webserver/auth.py [hunk 2]
```diff
@@ -51,42 +53,97 @@ class UnregisterCheck(InputSchema):
...
     expiration: timedelta | None = Field(
         None,
         description="Time delta from creation time to expiration. If None, then it does not expire.",
     )
 
     model_config = ConfigDict(
+        alias_generator=AliasGenerator(
+            validation_alias=to_camel,
+        ),
+        from_attributes=True,
```
Comment (sanderegg) [inline packages/models-library/src/models_library/api_schemas_webserver/auth.py:67]: why is this necessary? I thought that was only for DB models

File: packages/models-library/src/models_library/api_schemas_webserver/auth.py [hunk 3]
```diff
@@ -51,42 +53,97 @@ class UnregisterCheck(InputSchema):
...
+    )
+    api_base_url: HttpUrl
+    api_key: str
+    api_secret: str
+
+    model_config = ConfigDict(
+        alias_generator=AliasGenerator(
+            serialization_alias=to_camel,
+        ),
+        from_attributes=True,
```
Comment (sanderegg) [inline packages/models-library/src/models_library/api_schemas_webserver/auth.py:101]: same here

File: packages/models-library/src/models_library/api_schemas_webserver/auth.py [hunk 4]
```diff
@@ -51,42 +53,97 @@ class UnregisterCheck(InputSchema):
...
-    api_secret: str
+class ApiKeyGet(OutputSchema):
+    id: IDStr
+    display_name: Annotated[str, Field(..., min_length=3)]
 
     model_config = ConfigDict(
+        alias_generator=AliasGenerator(
+            serialization_alias=to_camel,
+        ),
         from_attributes=True,
```
Comment (sanderegg) [inline packages/models-library/src/models_library/api_schemas_webserver/auth.py:140]: and here?

**Review by sanderegg — APPROVED** [review]
thanks!

File: packages/models-library/src/models_library/api_schemas_webserver/auth.py (hunk 2 above)
Comment (giancarloromeo) [reply packages/models-library/src/models_library/api_schemas_webserver/auth.py:67]: Service layer models are `dataclass`es.

File: packages/models-library/src/models_library/api_schemas_webserver/auth.py (hunk 3 above)
Comment (giancarloromeo) [reply packages/models-library/src/models_library/api_schemas_webserver/auth.py:101]: See above.

File: packages/models-library/src/models_library/api_schemas_webserver/auth.py (hunk 4 above)
Comment (giancarloromeo) [reply packages/models-library/src/models_library/api_schemas_webserver/auth.py:140]: See above.

File: services/web/server/src/simcore_service_webserver/api_keys/_repository.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
+import logging
+from datetime import timedelta
+
+import sqlalchemy as sa
+from aiohttp import web
+from models_library.products import ProductName
+from models_library.users import UserID
+from simcore_postgres_database.models.api_keys import api_keys
+from simcore_postgres_database.utils_repos import transaction_context
+from simcore_service_webserver.api_keys._models import ApiKey
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/web/server/src/simcore_service_webserver/api_keys/_repository.py:10]: Sure, a vscode refactoring changed also these imports.

File: services/web/server/src/simcore_service_webserver/api_keys/_repository.py
[the remark this answers is not in the document]
Comment (matusdrobuliak66) [reply services/web/server/src/simcore_service_webserver/api_keys/_repository.py:1]: Our current naming convention in webserver (which we can change of course) is:
 - Controller --> _handlers.py
 - Service --> _api.py
 - Repository --> _db.py
 
This makes sense when you look at how some older modules were historically named, and all the new ones comply with this naming convention. Anyway, it's just about reaching an agreement so we can maintain a consistent naming style.

**Review by matusdrobuliak66 — APPROVED** [review]
Thanks a lot! Looks good. Lets just agree on the naming convention :+1:

File: services/web/server/src/simcore_service_webserver/api_keys/_repository.py
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/web/server/src/simcore_service_webserver/api_keys/_repository.py:1]: Sure, thanks both for clarification. The names need a change in my opinion because they don't reflect exactly what they represent. The main issue is another: dependencies statically imported. This generates too much coupling for my taste.

File: services/web/server/src/simcore_service_webserver/api_keys/_repository.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    app: web.Application,
+    connection: AsyncConnection | None = None,
+    *,
+    user_id: UserID,
+    product_name: ProductName,
+    api_key_id: str,
+) -> None:
+    async with transaction_context(get_asyncpg_engine(app), connection) as conn:
+        stmt = api_keys.delete().where(
+            (api_keys.c.id == int(api_key_id))  # NOTE: remove after migration to str
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/api_keys/_repository.py:167]: Q: is there a test that will make this fail after the migration? That would be ideal. A note in the code "might" get forgotten.

File: services/web/server/src/simcore_service_webserver/api_keys/_rest.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        expiration=new_api_key.expiration,
+        user_id=req_ctx.user_id,
+        product_name=req_ctx.product_name,
+    )
+
+    api_key = ApiKeyCreateResponse.model_validate(
+        {
+            **asdict(created_api_key),
+            "api_base_url": "http://localhost:8000",
+        }  # TODO: https://github.com/ITISFoundation/osparc-simcore/issues/6340 # @pcrespov
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/api_keys/_rest.py:61]: Q: was this forgotten?

File: services/web/server/tests/unit/with_dbs/01/test_api_keys_rpc.py [hunk 8]
```diff
@@ -95,21 +101,21 @@ async def rpc_client(
...
-async def test_api_key_get(
-    fake_user_api_keys: list[str],
+async def test_get_api_key(
+    fake_user_api_keys: list[ApiKey],
     rpc_client: RabbitMQRPCClient,
     osparc_product_name: ProductName,
     logged_user: UserInfoDict,
 ):
-    for api_key_name in fake_user_api_keys:
+    for api_key in fake_user_api_keys:
```
Comment (GitHK) [inline services/web/server/tests/unit/with_dbs/01/test_api_keys_rpc.py:110]: Since the RPC endpoints are usually called within our codebase. We have a convention that allows for reuse of code and ensures endpoint callas are always tested. Please refactor the RPC endpoint's test to use the following pattern, that we use for all other endpoints.

It works like this:
1. define a function to call the RPC endpoint inside the service_library
2. add the test in this module to it
3. everybody calling that shared function can profit form it's code being covered and avoids introducing code duplication and issue related to it.


You can see an example if you see how this endpoint is used. https://github.com/ITISFoundation/osparc-simcore/blob/6eb9b2efefad6259cf1e88e812bce311694fc37a/packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/dynamic_scheduler/services.py#L51

Comment (GitHK) [reply services/web/server/tests/unit/with_dbs/01/test_api_keys_rpc.py:110]: I would suggest to refactor all the code in this module with this pattern.

**Review by GitHK — COMMENTED** [review]
I have a few observations

File: services/web/server/src/simcore_service_webserver/api_keys/_repository.py (hunk 6 above)
Comment (giancarloromeo) [reply services/web/server/src/simcore_service_webserver/api_keys/_repository.py]: I created an issue for that and linked it to the code.

File: services/web/server/src/simcore_service_webserver/api_keys/_repository.py
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/web/server/src/simcore_service_webserver/api_keys/_repository.py:1]: @matusdrobuliak66 

```python
# _api.py
import _db

async def foo(
    app: web.Application,  # why the Service should know it's in a "Web" application?
    *,
    something: str
)
    _db.bar(app, something)   # how would you test the Service layer in isolation? (without runtime tricks like Moneypatching everything?)

# file: _db.py
async def bar(
    app: web.Application,
    *,
    something: str
)
    do_something_with(something)
```

Pythonically speaking, that is very clean and compact at module level (1 function per op)... but complicates how the layers are coupled. I would put interfaces to decouple stuff so that we don't pass so much things between the involved parts.

**Review by jsaq007 — APPROVED** [review]

**Review by pcrespov — APPROVED** [review]
